Forms: Add Feedback and Feedback_Field classes and start using them to display data

enejb opened the pull request:
This PR add a few of new classes that make dealing with form responses easier. 

Fixes [FORMS-231](https://linear.app/a8c/issue/FORMS-231/improve-handling-of-form-responses-with-new-feedback-classes)

Currently we don't have a structure way that we deal with Form Responses. This PR in an attempt to improve this. 
And also add more test to the code base. 
We are also attempting to deal 

## Proposed changes:
* Make it easier to deal with from responses by creating form response object  (`Feedback::get( $post_id )` or `Feedback::from_submission( $post_data, $form, $current_post, $current_page_number ) ` that works the same as before being persistent to the DB and after. 
* Update all the different place to use the new class. ( Endpoint, On submission, csv export, export personal data ) 
* In this PR we are not saving thew new format but has tests that show that reading the new format will be the same as saving when we save it. 
* This PR also tries to make it easier to write unit and integration tests.  


### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
Do the tests pass? 

Add a new form with all the fields. 
Save the form. Does it work as expected? 
Notice that you got the email as expected. 

Visit the form responses. This Should work as before. 

Export the form responses. Do they work like you would expected them to? 
Export the persona data visiting Tools > Export Personal Data and entering an email address. 
Notice that beside the field we also supply the IP address. Consent, Date and URL where the form was submitted from.

enejb committed:
Forms: Adds a new Form_Responses class

enejb committed:
Add tests for Form_Response::strip_tags method

lezama commented on projects/packages/forms/src/contact-form/class-form-response.php:
```diff
@@ -0,0 +1,806 @@
+	 *
+	 * @var string
+	 */
+	protected $entry_title = '';
```
do we need this?

enejb replied:
yes. so that we can display to the user on what page the form was created on.

lezama commented on projects/packages/forms/src/contact-form/class-form-response.php:
```diff
@@ -0,0 +1,806 @@
+	 *
+	 * @var string
+	 */
+	protected $subject = '';
```
what the subject would be?

enejb replied:
This is used by akismet. To know if things are spam or not

enejb committed:
Make endpoint rendering work

enejb committed:
Make for submission work with the new response data

enejb committed:
Add test for get_render_value

enejb committed:
Update has_file tests

enejb committed:
Add tests for get_render_api_value method

enejb committed:
rename for clarity

enejb committed:
Simplify things

enejb committed:
Fix rendering on submit

enejb committed:
Rename Form_Response => Feedback

enejb committed:
Rename Response_Field => Feedback_Field

enejb committed:
Implement CSV Export

enejb committed:
minor update test method name

enejb committed:
Minor fixes

enejb committed:
Add reprecation

enejb committed:
Update the code coverage

enejb committed:
minor: test fixes

enejb committed:
Introduce Feedback_Entry class

enejb committed:
split constructore up, refactor form

enejb committed:
Add Feedback_Author Class

enejb committed:
Fix phan issue

enejb committed:
Add tests for Feedback_Author

enejb committed:
Fix tests

enejb committed:
minor refactor

enejb committed:
Refactor legacy content parsing in Feedback class

Modularized the parse_legacy_content method by extracting logic into helper methods for splitting content, extracting values and lines, parsing JSON and array formats, and processing fields. This improves readability, maintainability, and separation of concerns in legacy feedback parsing.

enejb committed:
Refactor feedback field rendering and API output

Refactored Feedback_Field to support context-aware rendering via get_render_value($context), and made get_render_api_value() private. Updated Feedback to allow flexible compiled field output shapes and context. Adjusted endpoint and tests to use the new API, improving clarity and extensibility for field rendering in different contexts.

enejb committed:
Add context csv

enejb committed:
Add more context

enejb committed:
Fixing php phan issues

enejb committed:
Improve test coverage

enejb committed:
Fix typo in method name for author info retrieval

Renamed get_computer_author_info to get_computed_author_info in Feedback_Author to correct a typo and ensure consistency in method naming.

enejb committed:
Rename legacy feedback properties in Feedback class

Renamed $feedback_id and $feedback_title to $legacy_feedback_id and $legacy_feedback_title throughout the Feedback class to clarify their legacy status and indicate they are not used in the new feedback system. Updated all references, getters, and related logic accordingly.

enejb committed:
Add more Feedback tests

enejb committed:
Add form count test

enejb committed:
Move the test

enejb committed:
More tests and phan fixes

enejb committed:
update baseline

enejb committed:
Add get_value assertions to Feedback_Field tests

Extended the test_get_render_value method to include assertions for the get_value method, ensuring both scalar and array values are correctly returned by Feedback_Field.

enejb committed:
Add feedback field tests

enejb committed:
remove unused method

enejb committed:
Refactor file field type handling in feedback forms

Replaces is_file_field() with a more generic is_of_type() method in Feedback_Field, simplifying type checks. Updates legacy file upload handling in Feedback to explicitly set the field type to 'file'. Adjusts and expands related unit tests to cover new logic and ensure correct file field behavior.

enejb committed:
Don't reimplement strip_tags method

enejb committed:
Export personal Data update to use FeedBack Class

edanzer reviewed (commented):
- [x] Confirmed all the PHP tests pass
- [x] Tested an existing form and confirmed submission data is saved and rendered correctly. 
- [x] Added a new form with different kinds of fields, including file uploads, multiple choice and others, and confirm submission data is saved and rendered correctly. 
- [x] Confirmed that if Jetpack CRM is enabled, contacts are saved to CRM as expected.
- [x] Exported form responses as CSV and to Google sheets, and confirmed that works as expected. 

I'm hesitant to sign off on the code review piece of this because we're changing a lot, and I don't know the code around saving and working with submission data well enough to know if we'll introduce unanticipated issues with downstream data. To review the code, I'd really need to spend a lot of time digging into this part of our form code so I understand it thoroughly.

lezama committed:
Fix typos in docblocks

- Fix "lable" → "label" in Feedback_Field class
- Fix "recieved" → "received" in Feedback class
- Fix "invividual" → "individual" in Feedback_Field class
- Complete incomplete class docblock for Feedback_Field

enejb committed:
Making ajax work.

enejb committed:
Add a test for label|value format

enejb committed:
Add test for get_json_data
